compare/compare.c

```diff
@@ -0,0 +1,639 @@
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "image.h"
+#include "../libass/ass.h"
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <dirent.h>
+#include <string.h>
+
+
+#define FFMAX(a,b) ((a) > (b) ? (a) : (b))
+#define FFMIN(a,b) ((a) > (b) ? (b) : (a))
+
+static void blend_image(Image8 *frame, int32_t x0, int32_t y0,
+                        const ASS_Image *img)
+{
+    int32_t x1 = img->dst_x, x_min = FFMAX(x0, x1);
+    int32_t y1 = img->dst_y, y_min = FFMAX(y0, y1);
+    x0 = x_min - x0;  x1 = x_min - x1;
+    y0 = y_min - y0;  y1 = y_min - y1;
+
+    int32_t w = FFMIN(x0 + frame->width,  x1 + img->w);
+    int32_t h = FFMIN(y0 + frame->height, y1 + img->h);
+    if (w <= 0 || h <= 0)
+        return;
+
+    uint8_t r = img->color >> 24;
+    uint8_t g = img->color >> 16;
+    uint8_t b = img->color >>  8;
+    uint8_t a = img->color >>  0;
+
+    int32_t mul = 129 * (255 - a);
+    const int32_t offs = (int32_t) 1 << 22;
+
+    int32_t stride = 4 * frame->width;
+    uint8_t *dst = frame->buffer + y0 * stride + 4 * x0;
+    const uint8_t *src = img->bitmap + y1 * img->stride + x1;
+    for (int32_t y = 0; y < h; y++) {
+        for (int32_t x = 0; x < w; x++) {
+            int32_t k = src[x] * mul;
+            dst[4 * x + 0] -= ((dst[4 * x + 0] - r) * k + offs) >> 23;
+            dst[4 * x + 1] -= ((dst[4 * x + 1] - g) * k + offs) >> 23;
+            dst[4 * x + 2] -= ((dst[4 * x + 2] - b) * k + offs) >> 23;
+            dst[4 * x + 3] -= ((dst[4 * x + 3] - 0) * k + offs) >> 23;
+        }
+        dst += stride;
+        src += img->stride;
+    }
+}
+
+static void blend_all(Image8 *frame, int32_t x0, int32_t y0,
+                      const ASS_Image *img)
+{
+    uint8_t *dst = frame->buffer;
+    size_t size = (size_t) frame->width * frame->height;
+    for (size_t i = 0; i < size; i++) {
+        dst[0] = dst[1] = dst[2] = 0;
+        dst[3] = 255;
+        dst += 4;
+    }
+    for (; img; img = img->next)
+        blend_image(frame, x0, y0, img);
+}
+
+inline static uint16_t abs_diff(uint16_t a, uint16_t b)
+{
+    return a > b ? a - b : b - a;
+}
+
+inline static uint16_t abs_diff4(const uint16_t a[4], const uint16_t b[4])
+{
+    uint16_t res = 0;
+    for (int k = 0; k < 4; k++) {
+        uint16_t diff = abs_diff(a[k], b[k]);
+        res = FFMAX(res, diff);
+    }
+    return res;
+}
+
+static void calc_grad(const Image16 *target, uint16_t *grad)
```

Comment:
This needs some comments on what's being done here.

Reply:
I don't get what it does either. Some sort of fuzzy comparison?

Reply:
It's to account for spatial error.
For example, 1/4 pixel offset can transform `0 0 0 255 255` into `0 0 64 255 255`, therefore error of 64 can be acceptable in one pixel position but not others.
`calc_grad()` estimates value change around each pixel to account for that.
It's not the best metric but I think it can be starting point.